### PR TITLE
Module 3 React Group Project

### PR DESCRIPTION
**Pull request by mavericksbalitaan**
Hi there! 👋 

In this project, we (@mavericks-db, and @leonard33 ) build a web application for a company that provides commercial and scientific space travel services. The application will allow users to book rockets and join selected space missions.

Kindly check the live demo [here](https://app-space-travelers-hub.netlify.app/) 👈 

## Here are the features:
- [x] Config & basic setup
   - Create React App.
   - Install [React Redux](https://react-redux.js.org/), [Redux Logger](https://www.npmjs.com/package/redux-logger) and [React Router](https://reactrouter.com/web/guides/quick-start).
   - Download the [free image](https://www.flaticon.com/free-icon/planet_3212567?term=space&page=1&position=19&page=1&position=19&related_id=3212567&origin=style) for the app logo.
   - Create routes and view components (Rockets, Missions, My Profile). Use `<NavLink />` for the page navigation links and style active class to indicate which section/page the user is currently on (underline active navigation link).
   - Create directories for all Redux state slice files (rockets, missions).
- [x] Redux: Fetch data and update the Redux store
   - Upon first rendering fetch data from the SpaceX API endpoints:
      - Rockets: `https://api.spacexdata.com/v3/rockets`
      - Missions: `https://api.spacexdata.com/v3/missions`
   - Once the data are fetched, dispatch an action to store the selected data in the Redux store:
      - Rockets:
         - id
         - rocket_name
         - description
         - flickr_images
      - Missions:
         - mission_id
         - mission_name
         - description
NOTE: Make sure you only dispatch those actions once and do not add data to store on every re-render (i.e. when changing views / using navigation).

NOTE: Rockets are the default view, so you must fetch rockets data when the application starts. However, the mission data should only be fetched (once) when a user navigates to the Missions section.

 - [x] Render UI:lists
   - Use useSelector(), Redux Hook, to select the state slices and render lists of rockets and missions in corresponding routes. i.e.:
```
// get rockets data from the store
const rockets = useSelector(state => state.rockets);
```
   - You can style the whole application "by hand" or you could use [React Bootstrap](https://react-bootstrap.github.io/), a UI library that could speed up the process. This is a popular library and working with its components would be good practice.
   - Render a list of rockets (as per design). For the image of a rocket using the first image in the array of flickr_images.
   - Render a table with the missions' data (as per design).

- [x] Redux: Write actions and reducers for booking rockets/dragons and joining missions
   - When a user clicks the "Reserve rocket" button or "Reserve dragon" button [only if your team has 3 members], action needs to be dispatched to update the store. You need to get the ID of the reserved rocket and update the state. Remember you mustn't mutate the state. Instead, you need to return a new state object with all rockets, but the selected rocket will have an extra key `reserved` with its value set to `true`. You could use a JS `filter()` or `map()` to set the value of the new state - i.e.:
```
const newState = state.map(rocket => {
    if(rocket.id !== id) 
        return rocket;
    return { ...rocket, reserved: true };
});
```
   - Regardless of which method you choose, make sure you place all your logic in the reducer. In the React view file, you should only dispatch the action with the correct rocket ID as an argument.
   - Create a reducer and action dispatcher for the "Join Mission" button. The logic here is practically the same as with rockets - you need to pass the mission's ID to the corresponding action and update the mission's state with the selected mission having a new key/value - `reserved: true`.
 
- [x] Redux: Write actions and reducers for canceling rockets/dragons and leaving missions
  - Here you need to follow the same logic as with the "Reserve rocket"/"Reserve dragon" and "Join mission" - but you need to set the `reserved` key to `false`.
  - Dispatch these actions upon clicking on the corresponding buttons.

- [x] Render UI: conditional components rendering
  - Rockets that have already been reserved should show a "Reserved" badge and "Cancel reservation" button instead of the default "Reserve rocket" (as per design).
  - Missions that the user has joined already should show a badge "Active Member" instead of the default "NOT A MEMBER" and a button "Leave Mission" instead of the "Join Mission" button (as per design).
  - Rockets/Dragons and Missions should use the React conditional rendering syntax:
```
{rocket.reserved && ( 
    // render Cancel Rocket button
)}
```

- [x] Render UI: My Profile section
  - Compose two/three-column layout and list ONLY the rockets/dragons reserved and missions joined by the user (as per design):
     - Render a list of all joined missions (use `filter()`).
     - Render a list of all reserved rockets (use `filter()`).

- [x] Technical setup**
   - Set up the repository on GitHub and use Gitflow.
   - Set up React and Redux.
   - Set up testing libraries(React Testing Library and Jest).
   - Do not preserve user data.

### CODE REVIEW

File: src/components/rockets/Rockets.test.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+const MockRockets = () => (
+  <Provider store={store}>
+    <Rockets />
+  </Provider>
+);
+
+test('renders correctly', () => {
+  const tree = renderer.create(<MockRockets />).toJSON();
+  expect(tree).toMatchSnapshot();
+});
```
Comment (Gambit142) [inline src/components/rockets/Rockets.test.js]: - [x] Great work using react-test-renderer to test components. Kindly write tests for this component using React Testing Library just the same way you did for the `Mission` component. 👍

File: src/redux/missions/MissionsSlice.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        missionid: el.mission_id,
+        missionName: el.mission_name,
+        missionDesc: el.description,
+        missionweblink: el.wikipedia,
+        missionjoin: false,
+      }));
+      return missions;
+    },
+  },
+});
```
Comment (Gambit142) [inline src/redux/missions/MissionsSlice.js:50]: - [x] Kindly use `Jest` to write unit tests for your reducers and other pure functions. 👍

File: src/components/myprofile/Myprofile.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+              <button className="remBtn" type="button" id={el.rocketId} onClick={cancelHandler}>
+                Cancel
+              </button>
+            </div>
+          ))}
+        </div>
+      </div>
+    </section>
+  );
+}
```
Comment (Gambit142) [inline src/components/myprofile/Myprofile.js:69]: - [x] Great work implementing the optional features for this project, i.e leaving a mission and cancelling a reservation. The team did an amazing job here. Kudos 👏

**Review by Gambit142 — CHANGES_REQUESTED** [review]
Hi @mavericks-db @leonard33 👋, 

#### The team did a great job executing this project. Thank you for taking the time to implement the features correctly 👏 

## Highlights
- Installed all the necessary dependencies ✔️ 
- Setup React components correctly ✔️ 
- Configured redux store correctly ✔️
- Setup actions, action creators and reducers correctly ✔️ 
- Retrieved data from the store using `useSelector` hook ✔️

However, there are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Kindly check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:

N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

File: src/components/rockets/Rockets.test.js (hunk 1 above)
Comment (mavericksbalitaan) [reply src/components/rockets/Rockets.test.js]: Done in commit id: https://github.com/mavericks-db/space-travelers-hub/pull/35/commits/d5ce37d11ff0aab2e194e559312082dd650ea120

File: src/redux/missions/MissionsSlice.js (hunk 2 above)
Comment (mavericksbalitaan) [reply src/redux/missions/MissionsSlice.js:50]: Done in commit id: https://github.com/mavericks-db/space-travelers-hub/pull/35/commits/170572e42c8d85200f5639e9dfee855a94953941

File: src/components/myprofile/Myprofile.js (hunk 3 above)
Comment (mavericksbalitaan) [reply src/components/myprofile/Myprofile.js:69]: Thank you 👍

**Review by Mihndim2020 — APPROVED** [review]
# STATUS: APPROVED!!!  🎉 🎉 🎉 :champagne: :champagne: :champagne:
Hi Team, 

Your project is complete! Great work on making the changes requested by the previous reviewer. There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 🎉 🎉 :champagne: :champagne: :champagne:

Congratulations-Memes-FEATURED

## Optional suggestions

_Check the comments under the review._

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

### N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**